src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -0,0 +1,314 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import importlib.resources
+import json
+import logging
+from dataclasses import dataclass
+from typing import ClassVar, Iterable, Sequence, cast
+
+from pants.backend.python.target_types import UnrecognizedResolveNamesError
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.unions import UnionMembership, union
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    CoursierResolvedLockfile,
+)
+from pants.option.subsystem import Subsystem
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet
+
+DEFAULT_TOOL_LOCKFILE = "<default>"
+NO_TOOL_LOCKFILE = "<none>"
+
+
+logger = logging.getLogger(__name__)
+
+
+class JvmToolBase(Subsystem):
+    """Base class for subsystems that configure a set of artifact requirements for a JVM tool."""
+
+    # Default version of the tool. (Subclasses must set.)
+    default_version: ClassVar[str]
+
+    # Default artifacts for the tool in GROUP:NAME format. The `--version` value will be used for the
+    # artifact version if it has not been specified for a particular requirement. (Subclasses must set.)
+    default_artifacts: ClassVar[Sequence[str]]
```

Comment:
I'm a bit concerned that we'll need to deprecate or complicate this syntax if we add more features to `jvm_artifact` (excludes, etc)... it would be good to already know what the migration path from this syntax to specifying these arguments with a target will be.

It seems like it should be possible to support both simultaneously, but let's just start by assuming that we _will_ need to add target support, and that maybe this syntax stays around as the "easy mode" way to specify deps?

Reply:
Should we have a syntax for differentiating between the use cases? Or would we first try to interpret as an address and then fallback to interpreting as a coordinate?

Maybe we should just go right to addresses only to start?

Reply:
Based on discussion yesterday, I think that we'll always need a way to specify the builtin version, and even if we did have a "synthetic target" based approach, we'd still need options to seed the target. So this seems fine for now.

Reply:
> Maybe we should just go right to addresses only to start?

Maybe, yea? In the absence of synthetic targets though, that would amount to having two codepaths: one for builtin requirements, the other for targets. But probably not too much complexity, honestly.

Reply:
Hmm, would we be fine with deprecating the coordinate syntax later on?

Reply:
Yes, although obviously easier to do before stabilizing the JVM. Sorry to wobble back and forth, but I do sortof think that accepting addresses here would be straightforward... I think that you could just request classpath entries for them with `ClasspathEntryRequest.for_targets`, and it would automatically support either compiling a tool or downloading it.

Reply:
No worries. Will implement address support then, with fallback to coordinate syntax.

Reply:
Using `ClasspathEntryRequest.for_targets` though doesn't seem to generate coordinates needed to invoke Coursier though to obtain a lockfile. Instead it obtains actual jars to form the invokable classpath. Am I missing someting?

My thought to support addresses is to scan the transitive dependencies of the user-supplied addresses and extract `jvm_artifacts` and extract their coordinates. Not sure how this would play with `url=` option on those targets though.

Reply:
Oh, doh. Yes, would only be able to consume `jvm_artifacts` initially. Sorry about that.